### PR TITLE
WIP: Execute markdown after content loaded by later middlewares

### PR DESCRIPTION
**Pull request by jimjimovich**
### 1. What does this change do, exactly?
WIP for Issue #2139 
Continue the work which was started in #1611 and #1649 to make the staticfiles file server
the only middleware that hits the disk and loads content by extending this feature to markdown. This would allow the markdown directive to be used with other middlewares such as proxy.

### 2. Please link to the relevant issues.
 #2139, #1611, and #1649

### 3. Which documentation changes (if any) need to be made because of this PR?
Might want to add documentation noting that it markdown can be used with other directives like proxy.

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I am willing to help maintain this change if there are issues with it later

### 5. Issues
This is not ready to merge, or even functioning. I'm starting the pull request in order to show some progress and get help.  Specifically, line 109 of markdown.go causes a fatal error (segmentation violation). I copied this line from the work done on templates.go, and I just don't understand buffers enough to understand what I'm doing wrong.

### CODE REVIEW

**Comment by mholt** [comment]
Cool, thanks for working on this! I will be following your progress best I can.

(I wonder if you'd benefit from Blackfriday v2, which has better support for more advanced Markdown features. Depending on how you feel when this is done, I wonder if #1957 would be of interest to you.)

**Comment by jimjimovich** [comment]
I noticed that there were several Markdown related issues and feature requests. I would certainly be interested in helping out on those, but wanted to keep this pull request specifically focused on the linked two issues.  I also might have some more issues/feature requests to add and then work on before this is done.

For now, if someone could help me figure out what I'm doing wrong with the buffers on line 109, that would be a huge help!

**Comment by mholt** [comment]
Cool -- yeah, definitely one (significant) change per pull request is preferred.

What's wrong the buffers, exactly?

**Comment by jimjimovich** [comment]
I get this error when accessing a page that uses the markdown directive. I've narrowed it down to line [109](https://github.com/mholt/caddy/pull/2143/commits/6f238ac9cdab04ec701c516ee49f2a3f3e84c182#diff-4adf3d16b980e337005e829c6541c23cR109) where the buffer code starts.

```fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x471a0b]

goroutine 7 [running]:
runtime.throw(0xb90978, 0x2a)
        /opt/go/src/runtime/panic.go:616 +0x81 fp=0xc42007c750 sp=0xc42007c730 pc=0x42c691
runtime.sigpanic()
        /opt/go/src/runtime/signal_unix.go:372 +0x28e fp=0xc42007c7a0 sp=0xc42007c750 pc=0x441f1e
sync.(*Pool).pin(0x0, 0xc)
        /opt/go/src/sync/pool.go:186 +0x2b fp=0xc42007c7c0 sp=0xc42007c7a0 pc=0x471a0b
sync.(*Pool).Get(0x0, 0x16, 0xc42003c300)
        /opt/go/src/sync/pool.go:128 +0x2f fp=0xc42007c808 sp=0xc42007c7c0 pc=0x47173f
github.com/mholt/caddy/caddyhttp/markdown.Markdown.ServeHTTP(0xc42003a300, 0x2b, 0xc00fe0, 0xc42019c560, 0xc00a80, 0xc42024b890, 0xc42019c550, 0x2, 0x2, 0x0, ...)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddyhttp/markdown/markdown.go:109 +0x161 fp=0xc42007c9f0 sp=0xc42007c808 pc=0x9a2631
github.com/mholt/caddy/caddyhttp/markdown.(*Markdown).ServeHTTP(0xc420190910, 0xc065a0, 0xc4202a8000, 0xc420286500, 0x0, 0xc42007e300, 0xc06ea0)
        <autogenerated>:1 +0x8b fp=0xc42007ca80 sp=0xc42007c9f0 pc=0x9a915b
github.com/mholt/caddy/caddyhttp/httpserver.(*Server).serveHTTP(0xc4200b0420, 0xc065a0, 0xc4202a8000, 0xc420286500, 0x5, 0xc420030f00, 0xc06ea0)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:434 +0x2e0 fp=0xc42007cbf0 sp=0xc42007ca80 pc=0x9253b0
github.com/mholt/caddy/caddyhttp/httpserver.(*Server).ServeHTTP(0xc4200b0420, 0xc065a0, 0xc4202a8000, 0xc420286400)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:370 +0x2cf fp=0xc42007cd58 sp=0xc42007cbf0 pc=0x924faf
net/http.serverHandler.ServeHTTP(0xc42028c820, 0xc065a0, 0xc4202a8000, 0xc420286100)
        /opt/go/src/net/http/server.go:2694 +0xbc fp=0xc42007cd88 sp=0xc42007cd58 pc=0x6b6b9c
net/http.(*conn).serve(0xc4202741e0, 0xc06de0, 0xc420046980)
        /opt/go/src/net/http/server.go:1830 +0x651 fp=0xc42007cfc8 sp=0xc42007cd88 pc=0x6b2f11
runtime.goexit()
        /opt/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42007cfd0 sp=0xc42007cfc8 pc=0x45b961
created by net/http.(*Server).Serve
        /opt/go/src/net/http/server.go:2795 +0x27b

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc42003c16c)
        /opt/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc42003c160)
        /opt/go/src/sync/waitgroup.go:129 +0x72
github.com/mholt/caddy.(*Instance).Wait(0xc42026c000)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy.go:438 +0x2f
github.com/mholt/caddy/caddy/caddymain.Run()
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy/caddymain/run.go:147 +0x3e3
main.main()
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy/main.go:27 +0x27

goroutine 35 [syscall]:
os/signal.signal_recv(0x0)
        /opt/go/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
        /opt/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
        /opt/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 36 [select]:
github.com/mholt/caddy/caddytls.maintainAssets(0xc4200fe120)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddytls/maintain.go:68 +0x155
created by github.com/mholt/caddy/caddytls.init.2
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddytls/maintain.go:32 +0x4e

goroutine 3 [chan receive]:
github.com/mholt/caddy.trapSignalsCrossPlatform.func1()
        /home/ubuntu/workspace/src/github.com/mholt/caddy/sigtrap.go:43 +0x100
created by github.com/mholt/caddy.trapSignalsCrossPlatform
        /home/ubuntu/workspace/src/github.com/mholt/caddy/sigtrap.go:38 +0x35

goroutine 4 [chan receive]:
github.com/mholt/caddy.trapSignalsPosix.func1()
        /home/ubuntu/workspace/src/github.com/mholt/caddy/sigtrap_posix.go:32 +0x13f
created by github.com/mholt/caddy.trapSignalsPosix
        /home/ubuntu/workspace/src/github.com/mholt/caddy/sigtrap_posix.go:28 +0x35

goroutine 20 [select, locked to thread]:
runtime.gopark(0xbaa020, 0x0, 0xb7601c, 0x6, 0x18, 0x1)
        /opt/go/src/runtime/proc.go:291 +0x11a
runtime.selectgo(0xc4200d7750, 0xc4200b20c0)
        /opt/go/src/runtime/select.go:392 +0xe50
runtime.ensureSigM.func1()
        /opt/go/src/runtime/signal_unix.go:549 +0x1f4
runtime.goexit()
        /opt/go/src/runtime/asm_amd64.s:2361 +0x1

goroutine 6 [IO wait]:
internal/poll.runtime_pollWait(0x7f04b862cf00, 0x72, 0x0)
        /opt/go/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420250718, 0x72, 0xc420046800, 0x0, 0x0)
        /opt/go/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc420250718, 0xffffffffffffff00, 0x0, 0x0)
        /opt/go/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Accept(0xc420250700, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /opt/go/src/internal/poll/fd_unix.go:372 +0x1a8
net.(*netFD).accept(0xc420250700, 0xc420274260, 0xc42007bdd8, 0x402a98)
        /opt/go/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc420194110, 0xc42007be08, 0x4019c7, 0xc420274260)
        /opt/go/src/net/tcpsock_posix.go:136 +0x2e
net.(*TCPListener).AcceptTCP(0xc420194110, 0xc42007be50, 0xc42007be58, 0x18)
        /opt/go/src/net/tcpsock.go:246 +0x49
github.com/mholt/caddy/caddyhttp/httpserver.tcpKeepAliveListener.Accept(0xc420194110, 0xba9998, 0xc4202741e0, 0xc06ea0, 0xc420030cf0)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:517 +0x2f
net/http.(*Server).Serve(0xc42028c820, 0xc06820, 0xc420194110, 0x0, 0x0)
        /opt/go/src/net/http/server.go:2770 +0x1a5
github.com/mholt/caddy/caddyhttp/httpserver.(*Server).Serve(0xc4200b0420, 0xc06820, 0xc420194110, 0x0, 0x0)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddyhttp/httpserver/server.go:321 +0xa6
github.com/mholt/caddy.startServers.func1.1(0xc4200b0480, 0xc07060, 0xc4200b0420, 0xc06820, 0xc420194110, 0xc42026c000)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy.go:784 +0x49
created by github.com/mholt/caddy.startServers.func1
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy.go:783 +0xac

goroutine 25 [chan receive]:
github.com/mholt/caddy.startServers.func2(0xc4200b0480)
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy.go:796 +0x41
created by github.com/mholt/caddy.startServers
        /home/ubuntu/workspace/src/github.com/mholt/caddy/caddy.go:795 +0xb27

goroutine 8 [runnable]:
net/http.(*connReader).backgroundRead(0xc420030de0)
        /opt/go/src/net/http/server.go:667
created by net/http.(*connReader).startBackgroundRead
        /opt/go/src/net/http/server.go:664 +0xce
exit status 2
```

**Comment by mholt** [comment]
I think your buffer pool is nil? https://github.com/mholt/caddy/pull/2143/files#diff-4adf3d16b980e337005e829c6541c23cR50

It needs to be initialized: https://golang.org/pkg/sync/#example_Pool

**Comment by jimjimovich** [comment]
@mholt Thanks! That was exactly the problem. I didn't see that line in the templates setup where I was copying ideas/code from.

**Comment by jimjimovich** [comment]
@mholt Just a quick question about the process for squashing these commits. Is that possible to do after finishing up the work on this PR or did I need to do that before I made a PR?

**Comment by mholt** [comment]
@jimjimovich I should update the template, that was before GitHub allowed us to merge PRs by squashing. I can squash on this end. But, if there are multiple significant changes in a PR (which I discourage anyways), squashing each individual one to a single commit can be useful, and still keeping the multiple commits, in case we need to revert one later. This change should probably just be one commit, which I can squash.

**Comment by jimjimovich** [comment]
I updated the code. I added some TODO lines where I need help. If someone could take a look at this, I could use some help.
Thanks!

**Comment by jimjimovich** [comment]
I've been using these test files and Caddyfile to test things.
https://github.com/jimjimovich/markdown-test

File: caddyhttp/markdown/markdown.go [hunk 1]
```diff
@@ -43,6 +45,8 @@ type Markdown struct {
 
 	// The list of markdown configurations
 	Configs []*Config
+
+	BufPool *sync.Pool // docs: "A Pool must not be copied after first use."
```
Comment (tobya) [inline caddyhttp/markdown/markdown.go]: I dont think we need this comment, you can remove it.

File: caddyhttp/markdown/markdown.go [hunk 2]
```diff
@@ -97,94 +101,81 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
...
+	w.WriteHeader(http.StatusOK)
 	if r.Method == http.MethodGet {
 		w.Write(html)
 	}
-	return http.StatusOK, nil
+
+	return 0, nil
 }
 
+// TODO: This is not currently used in new implementation, do we need it?
```
Comment (tobya) [inline caddyhttp/markdown/markdown.go]: If its not used take it out rather than commenting it out.

File: caddyhttp/markdown/markdown.go [hunk 3]
```diff
@@ -97,94 +101,81 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 		return http.StatusMethodNotAllowed, nil
 	}
 
+	// TODO: Deal with dirents variable in new implementation
 	var dirents []os.FileInfo
-	var lastModTime time.Time
-	fpath := r.URL.Path
-	if idx, ok := httpserver.IndexFile(md.FileSys, fpath, cfg.IndexFiles); ok {
-		// We're serving a directory index file, which may be a markdown
```
Comment (tobya) [inline caddyhttp/markdown/markdown.go]: Where has all this logic gone?  Is it no longer needed?

File: caddyhttp/markdown/markdown.go (hunk 1 above)
Comment (jimjimovich) [reply caddyhttp/markdown/markdown.go]: I copied this line (along with the comment) from the template implementation from #1649 
https://github.com/mholt/caddy/blob/2922d09bef3c504dde66bc12f7441668fcef6a20/caddyhttp/templates/templates.go#L132

I also don't think the comment needs to be there, but I'll defer to someone with more knowledge of the situation.

File: caddyhttp/markdown/markdown.go (hunk 2 above)
Comment (jimjimovich) [reply caddyhttp/markdown/markdown.go]: Yes, I'd like to take it out, but would like someone to review if it's still needed or not before taking it out.

File: caddyhttp/markdown/markdown.go (hunk 3 above)
Comment (jimjimovich) [reply caddyhttp/markdown/markdown.go]: I believe this functionality is handled by staticfiles.

**Comment by jimjimovich** [comment]
Seems that something broke some 301 redirects when templates was updated in a similar way in #1649. The same thing is happening here. We should not merge this until we solve the issue with redirects as described here #2114

**Comment by jimjimovich** [comment]
I'm planning to continue working on this and writing more tests. There are still quite a few areas that need tests. I also want to take some time to fully understand the whole code base. 

One thing I'm kind of confused about is the dirents variable and several references to "files." I don't see how it was ever possible to have more than one markdown file or more than one template file. Any insight into this that someone could give me would be very helpful.

**Comment by mholt** [comment]
Sounds good; thanks for all your hard work on this. Take the time you need! If you need people to help contribute, couldn't hurt to let people know.

> One thing I'm kind of confused about is the dirents variable and several references to "files." I don't see how it was ever possible to have more than one markdown file or more than one template file. Any insight into this that someone could give me would be very helpful.

I'm not totally sure what you mean here? Maybe make a question inline with the code.

File: caddyhttp/markdown/markdown.go (hunk 3 above)
Comment (jimjimovich) [reply caddyhttp/markdown/markdown.go]: While staticfiles seems to handle the standard situation of index.html or other index files that have been configured outside of the markdown directive, the existing markdown code does have a config of IndexFiles that includes index.md, index.markdown, and index.mdown by default and would have something like index.ext extensions were set in the markdown configuration.

Is there a way to send this list of possible index files up the change when calling this?
https://github.com/jimjimovich/caddy/blob/b0f4085af3f02d289144c6a84e739109cdf00ae3/caddyhttp/markdown/markdown.go#L135-L136

Comment (jimjimovich) [reply caddyhttp/markdown/markdown.go]: Perhaps having index files defined in the markdown plugin is actually a bad idea and they should instead be defined in http.index?

https://caddyserver.com/docs/index

Comment (mholt) [reply caddyhttp/markdown/markdown.go]: Yes, definitely let the staticfiles package figure out index files. :+1:

File: caddyhttp/markdown/setup.go [hunk 4]
```diff
@@ -41,6 +41,11 @@ func setup(c *caddy.Controller) error {
 
 	cfg := httpserver.GetConfig(c)
 
+	// Add markdown index files to *SiteConfig.IndexPages
+	for _, mdc := range mdconfigs {
+		cfg.IndexPages = append(cfg.IndexPages, mdc.IndexFiles...)
```
Comment (jimjimovich) [inline caddyhttp/markdown/setup.go]: If someone could help to de-duplicate the resulting string, that would be a great help.  

With this implementation, the resulting IndexPages can contain many duplicates like this: [index.html index.htm index.txt default.html default.htm default.txt index.html index.txt index.md index.txt index.md index.html index.json]

Comment (mholt) [reply caddyhttp/markdown/setup.go]: Or just iterate the list (it's usually small, less than ~30 items) and see if it exists, and if not, then add it.

File: caddyhttp/markdown/markdown.go [hunk 5]
```diff
@@ -90,101 +94,81 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
...
 
+	// copy the buffered header into the real ResponseWriter
+	rb.CopyHeader()
+
 	w.Header().Set("Content-Type", "text/html; charset=utf-8")
 	w.Header().Set("Content-Length", strconv.Itoa(len(html)))
-	httpserver.SetLastModifiedHeader(w, lastModTime)
+	w.WriteHeader(http.StatusOK)
 	if r.Method == http.MethodGet {
 		w.Write(html)
```
Comment (jimjimovich) [inline caddyhttp/markdown/markdown.go]: What is better here, to have 
```
w.WriteHeader(http.StatusOK)
w.Write(html)
```
or 

```
http.ServeContent(rb.StatusCodeWriter(w), r, fpath, lastModTime, bytes.NewReader(html))
```

Comment (mholt) [reply caddyhttp/markdown/markdown.go]: The ServeContent one. ServeContent ensures that Range headers are honored, as well as ETags.

**Comment by jimjimovich** [comment]
I made quite a bit of progress, but there is still more to do. More tests need to be written and I need to make sure that things are working well with requests other than GET and HEAD.  Not sure if I'll work on this on the weekend or not, but hope to finish this up early next week if possible.

File: caddyhttp/markdown/setup.go (hunk 4 above)
Comment (jimjimovich) [reply caddyhttp/markdown/setup.go]: @Mohammed90 Unfortunately, converting it to a map makes it lose the order of files, which in this case is important.

File: caddyhttp/markdown/setup.go [hunk 6]
```diff
@@ -39,10 +41,30 @@ func setup(c *caddy.Controller) error {
...
+				if indexFile == indexPage {
+					inIndexPages = true
+				}
+			}
+			if inIndexPages == false {
+				cfg.IndexPages = append(cfg.IndexPages, indexFile)
+			}
+		}
+	}
+
```
Comment (jimjimovich) [inline caddyhttp/markdown/setup.go]: @mholt I think you were suggesting doing something like this. This code gets the correct result into cfg.IndexPages without duplicates, but being a bit new to Go, I'm not sure this is the best way as it looks a bit messy.

Comment (mholt) [reply caddyhttp/markdown/setup.go]: That works! It's clear and easy to read.

**Comment by jimjimovich** [comment]
@mholt and others,
I think this is ready for review. I'm not sure if it needs more tests or not, I added one for the 301 redirect fix, but most of the other cases seemed to be covered by existing tests.

**Comment by jimjimovich** [comment]
Also, the Summarize function would need to be updated to not look for a file but to use the buffer. However, I can't seem to see where this is ever used. I also don't see anything about summaries in the documentation. 

https://github.com/mholt/caddy/blob/148a6f443075e4a527ee0926279ee52f657e616b/caddyhttp/markdown/process.go#L41-L56

**Comment by jimjimovich** [comment]
Also, we need to deal with HEAD requests as they are returning a 0 content length and in some cases just not working. It seems like we need to actually process the markdown to get the real content in order to get a content length and etag and not rely on what staticfiles is sending back.

**Comment by jimjimovich** [comment]
Actually, HEAD is working (I had a problem in my Caddyfile). The etag is the same for HEAD and GET, but the content-length is set to 0 for HEAD. Is that okay or do we need to process the Markdown and get the real content-length?

**Comment by jimjimovich** [comment]
I added a fix to temporarily change the HTTP method from HEAD to GET before sending the request to staticfiles, then change it back to HEAD.  That gives us the correct content-length header for a HEAD request.

Actually ready for a review now. Not planning to work on this for a few hours.

**Comment by jimjimovich** [comment]
I'm also wondering if we shouldn't re-calculate etags in markdown since the etag from staticfiles only takes into account the original markdown file. I'm not sure how to go about that, but it might be nice ... or might be not necessary.

**Comment by mholt** [comment]
Closing in favor of Caddy 2, as discussed elsewhere. :) Thanks for contributing! (Experience from this review process made its way into Caddy 2.)

**Pull request closed without merging**